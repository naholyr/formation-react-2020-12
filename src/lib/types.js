import type { State as TodozState } from "../components/todoz/reducer";

export type AppState = {
  todoz: TodozState,
  counterz: any,
};
