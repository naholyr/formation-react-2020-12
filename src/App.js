import logo from "./logo.svg";
import "./App.scss";
import BtnCounterList from "./BtnCounterList";
import { v4 as uuid } from "uuid";
import { useCallback, useState } from "react";
import BtnCounterAdd from "./BtnCounterAdd";

function App() {
  const [counterIds, setCounterIds] = useState([]);
  const [toggled, setToggled] = useState(false);

  const addCounter = useCallback(() => {
    setCounterIds([uuid(), ...counterIds]);
  }, [counterIds]);

  return (
    <div className="App">
      <h1>Awesome counters</h1>
      <span onClick={() => setToggled(!toggled)}>{toggled ? "Yes" : "No"}</span>
      <BtnCounterAdd onAddCounter={addCounter} />
      <BtnCounterList counterIds={counterIds} />
    </div>
  );
}

export default App;
