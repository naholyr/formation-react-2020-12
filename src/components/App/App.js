import React, { memo } from "react";
import Header from "../nav/Header/Header";
import MainContent from "../MainContent/MainContent";
import { useState } from "react";
import NavItemBadgeCounterz from "../nav/NavItem/NavItemBadgeCounterz";
import NavItemBadgeTodoz from "../nav/NavItem/NavItemBadgeTodoz";

const App = () => {
  const [currentPageId, setCurrentPageId] = useState("todoz");

  const [navItems] = useState(() => [
    { label: "Home", pageId: "home" },
    {
      label: "Counterz",
      pageId: "counterz",
      badgeLabel: <NavItemBadgeCounterz />,
    },
    { label: "Todoz", pageId: "todoz", badgeLabel: <NavItemBadgeTodoz /> },
  ]);

  return (
    <>
      <Header
        navItems={navItems}
        currentPageId={currentPageId}
        onClickNavItem={setCurrentPageId}
      />
      <MainContent currentPageId={currentPageId} />
    </>
  );
};

export default memo(App);
