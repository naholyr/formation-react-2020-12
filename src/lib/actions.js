export const addTodo = (text) => ({
  type: "TODOZ_ADD",
  payload: { text },
});
