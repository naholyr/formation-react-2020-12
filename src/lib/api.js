/*
const foo = async () => {
  const v = await p;
  console.log(v)
  return "toto"
}

const foo = () => {
  return p.then(v => {
    console.log(v)
    return "toto"
  })
}
*/
import { v4 as uuid } from "uuid";

export const fetchTasks = async () => {
  console.log("real fetchTasks");
  await delay(4000);
  const json = localStorage.getItem("todoz") || "[]";
  return JSON.parse(json);
};

export const addTask = async (label, tagLabels) => {
  await delay(3000);
  const object = { label, tagLabels, done: false, id: uuid() };

  // add to local storage
  const json = localStorage.getItem("todoz") || "[]";
  const tasks = JSON.parse(json);
  localStorage.setItem("todoz", JSON.stringify([...tasks, object]));

  return object;
};

export const removeTask = () => {
  throw new Error("NOT IMPLEMENTED");
};

export const updateTask = () => {
  throw new Error("NOT IMPLEMENTED");
};

const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
