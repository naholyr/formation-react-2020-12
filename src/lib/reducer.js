import { combineReducers } from "redux";
import { reducer as counterzReducer } from "../components/counterz/reducer";
import { reducer as todozReducer } from "../components/todoz/reducer";

export const reducer = combineReducers({
  todoz: todozReducer,
  counterz: counterzReducer,
});
