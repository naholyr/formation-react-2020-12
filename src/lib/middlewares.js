import * as api from "./api";

const logMiddleware = (store) => (next) => (action) => {
  console.log(action);
  next(action);
};

const titleMiddleware = (store) => (next) => (action) => {
  if (action.type === "SET_TITLE") {
    document.title = action.payload;
    // black hole
  } else {
    next(action);
  }
};

// Version générique
// Cf. redux-axios-middleware
// cf. redux-thunk
const asyncMiddleware = (store) => (next) => (action) => {
  if (action.type === action.async) {
    const load = async () => {
      store.dispatch({ type: action.payload.type + "_START" });
      try {
        const payload = await action.payload();
        store.dispatch({
          type: action.payload.type + "_SUCCESS",
          payload,
        });
      } catch (err) {
        store.dispatch({
          type: action.payload.type + "_ERROR",
          payload: { message: err.message },
        });
      }
    };
    load();
  } else {
    next(action);
  }
};

const fetchTasksMiddleware = (store) => (next) => (action) => {
  if (action.type === "TODOZ_FETCH") {
    const load = async () => {
      store.dispatch({ type: "TODOZ_FETCH_START" });
      try {
        const tasks = await api.fetchTasks();
        store.dispatch({
          type: "TODOZ_FETCH_SUCCESS",
          payload: { tasks },
        });
      } catch (err) {
        store.dispatch({
          type: "TODOZ_FETCH_ERROR",
          payload: { message: err.message },
        });
      }
    };
    load();
  } else {
    next(action);
  }
};

export const middlewares = [
  logMiddleware,
  titleMiddleware,
  fetchTasksMiddleware,
  asyncMiddleware,
];
