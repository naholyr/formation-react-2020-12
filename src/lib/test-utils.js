import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { applyMiddleware, createStore } from "redux";
import { middlewares } from "./middlewares";
import { reducer } from "./reducer";

export const renderWithStore = (element) => {
  const store = createStore(reducer, applyMiddleware(...middlewares));
  return render(<Provider store={store}>{element}</Provider>);
};
