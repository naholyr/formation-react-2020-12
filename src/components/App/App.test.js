import { screen } from "@testing-library/react";
import { renderWithStore } from "../../lib/test-utils";
import App from "./App";

test("renders learn react link", () => {
  renderWithStore(<App />);
  const [link, title] = screen.getAllByText(/Todoz/i);
  expect(link).toHaveAttribute("href", "#todoz");
  // TODO test title is H1 ??
});

test("App snapshot", () => {
  const { container } = renderWithStore(<App />);
  expect(container).toMatchSnapshot();
});
