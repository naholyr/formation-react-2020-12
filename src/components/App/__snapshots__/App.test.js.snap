// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div>
  <nav
    class="navbar navbar-expand navbar-light bg-light"
  >
    <button
      class="navbar-brand"
    >
      Awesome App
    </button>
    <div
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav mr-auto"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="#home"
          >
            Home
            
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="#counterz"
          >
            Counterz
            <span
              class="ml-1 badge badge-pill badge-primary"
            >
              0
            </span>
          </a>
        </li>
        <li
          class="nav-item active"
        >
          <a
            class="nav-link"
            href="#todoz"
          >
            Todoz
            <span
              class="ml-1 badge badge-pill badge-primary"
            >
              0/0
            </span>
            <span
              class="sr-only"
            >
              (current)
            </span>
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    class="container page"
  >
    <h1>
      Todoz
    </h1>
    <p>
      Loading…
    </p>
  </div>
</div>
`;

exports[`renders learn react link 1`] = `
<div>
  <nav
    class="navbar navbar-expand navbar-light bg-light"
  >
    <button
      class="navbar-brand"
    >
      Awesome App
    </button>
    <div
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav mr-auto"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="#home"
          >
            Home
            
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="#counterz"
          >
            Counterz
            <span
              class="ml-1 badge badge-pill badge-primary"
            >
              0
            </span>
          </a>
        </li>
        <li
          class="nav-item active"
        >
          <a
            class="nav-link"
            href="#todoz"
          >
            Todoz
            <span
              class="ml-1 badge badge-pill badge-primary"
            >
              0/0
            </span>
            <span
              class="sr-only"
            >
              (current)
            </span>
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    class="container page"
  >
    <h1>
      Todoz
    </h1>
    <p>
      Loading…
    </p>
  </div>
</div>
`;
